pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -97,26 +100,63 @@ func (pl *DefaultPreemption) PostFilter(ctx context.Context, state *framework.Cy
 // other pods with the same priority. The nominated pod prevents other pods from
 // using the nominated resources and the nominated pod could take a long time
 // before it is retried after many other pending pods.
-func preempt(ctx context.Context, fh framework.FrameworkHandle, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, error) {
-	cs := fh.ClientSet()
+func (pl *DefaultPreemption) preempt(ctx context.Context, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, error) {
+	cs := pl.fh.ClientSet()
+	ph := pl.fh.PreemptHandle()
+	nodeLister := pl.fh.SnapshotSharedLister().NodeInfos()
+
+	// 0) Fetch the latest version of <pod>.
 	// TODO(Huang-Wei): get pod from informer cache instead of API server.
 	pod, err := util.GetUpdatedPod(cs, pod)
 	if err != nil {
 		klog.Errorf("Error getting the updated preemptor pod object: %v", err)
 		return "", err
 	}
 
-	if !podEligibleToPreemptOthers(pod, fh.SnapshotSharedLister().NodeInfos(), m[pod.Status.NominatedNodeName]) {
-		klog.V(5).Infof("Pod %v/%v is not eligible for more preemption.", pod.Namespace, pod.Name)
-		return "", nil
+	// 1) Find all preemption candidates.
+	candidates, err := FindCandidates(ctx, cs, state, pod, m, ph, nodeLister, pl.pdbLister)
+	if err != nil || len(candidates) == 0 {
+		return "", err
 	}
-	allNodes, err := fh.SnapshotSharedLister().NodeInfos().List()
+
+	// 2) Interact with registered Extenders to filter out some candidates if needed.
+	candidates, err = CallPreemptExtenders(ph.Extenders(), pod, nodeLister, candidates)
 	if err != nil {
 		return "", err
 	}
+
+	// 3) Find the best candidate.
+	bestCandidate := SelectBestCandidate(candidates)
+	if bestCandidate == nil || len(bestCandidate.Name()) == 0 {
+		return "", nil
+	}
+
+	// 4) Apply the candidate winner.
+	if err := ApplyCandidate(bestCandidate, pl.fh, cs, pod); err != nil {
+		return "", err
+	}
+
+	return bestCandidate.Name(), nil
+}
+
+// FindCandidates calculates a slice of preemption candidates.
+// Each candidate is executable to make the given <pod> schedulable.
+func FindCandidates(ctx context.Context, cs kubernetes.Interface, state *framework.CycleState, pod *v1.Pod,
+	m framework.NodeToStatusMap, ph framework.PreemptHandle, nodeLister framework.NodeInfoLister,
+	pdbLister policylisters.PodDisruptionBudgetLister) ([]Candidate, error) {
+	if !podEligibleToPreemptOthers(pod, nodeLister, m[pod.Status.NominatedNodeName]) {
```

Comment:
I think In terms of logic keeping the `podEligibleToPreemptOthers judgement` in `func preempt` is better than into `FindCandidates`.  WDYT?

Reply:
SGTM. Users may customize `podEligibleToPreemptOthers` as well.